File: src/common/platform_set.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type t = Bitset.t [@@deriving show]

let available_platforms_to_bitset ~multi_platform_extensions available_platforms =
  let bitset = Bitset.all_zero (Base.List.length multi_platform_extensions) in
  Base.List.foldi multi_platform_extensions ~init:bitset ~f:(fun i bitset ext ->
      if Base.List.exists available_platforms ~f:(fun p -> ext = "." ^ p) then
        Bitset.set i bitset
      else
        bitset
  )

let available_platforms ~file_options ~filename ~explicit_available_platforms : t option =
  if not (Files.multi_platform file_options) then
    None
  else
    let multi_platform_extensions = Files.multi_platform_extensions file_options in
    match Files.platform_specific_extensions_and_indices_opt ~options:file_options filename with
    | Some result ->
      let bitset = Bitset.all_zero (Base.List.length multi_platform_extensions) in
      let bitset =
        Base.List.fold result ~init:bitset ~f:(fun bitset (i, _) -> Bitset.set i bitset)
      in
      Some bitset
    | None ->
      (match explicit_available_platforms with
      | None ->
        let match_directory_override (path, platform) =
          let normalized_filename = Sys_utils.normalize_filename_dir_sep filename in
          Base.Option.some_if (Base.String.is_prefix ~prefix:path normalized_filename) platform
        in
        (match
           Base.List.find_map
             (Files.multi_platform_ambient_supports_platform_directory_overrides file_options)
             ~f:match_directory_override
         with
        | None -> Some (Bitset.all_one (Base.List.length multi_platform_extensions))
        | Some supported_platforms_from_directory_overrides ->
          Some
            (available_platforms_to_bitset
               ~multi_platform_extensions
               supported_platforms_from_directory_overrides
            ))
      | Some explicit_available_platforms ->
        Some (available_platforms_to_bitset ~multi_platform_extensions explicit_available_platforms))

let is_subset = Bitset.is_subset

let no_overlap = Bitset.no_overlap

let to_platform_string_set ~file_options bitset =
  Base.List.foldi
    (Files.multi_platform_extensions file_options)
    ~init:SSet.empty
    ~f:(fun i acc ext ->
      if Bitset.mem i bitset then
        SSet.add (Base.String.chop_prefix_exn ~prefix:"." ext) acc
      else
        acc
  )

let platform_specific_implementation_mrefs_of_possibly_interface_file
    ~file_options ~platform_set ~file =
  if Files.multi_platform file_options && Files.has_flow_ext file then
    let file = Files.chop_flow_ext file in
    let platform_set = Base.Option.value_exn platform_set in
    Base.List.find_map (Files.module_file_exts file_options) ~f:(fun module_file_ext ->
        if File_key.check_suffix file module_file_ext then
          let base =
            File_key.chop_suffix file module_file_ext |> File_key.to_string |> Filename.basename
          in
          let platform_extensions =
            Base.List.filter_mapi
              (Files.multi_platform_extensions file_options)
              ~f:(fun i platform_ext ->
                if Bitset.mem i platform_set then
                  Some platform_ext
                else
                  None
            )
          in
          let multi_platform_extension_group_mapping =
            Files.multi_platform_extension_group_mapping file_options
          in
          let (unconditional_extensions, grouped_extensions_with_conditional_extensions) =
            Base.List.fold
              platform_extensions
              ~init:(SSet.empty, SMap.empty)
              ~f:(fun (unconditional_extensions, grouped_extensions_with_conditional_extensions) ext
                 ->
                match
                  Base.List.find
                    multi_platform_extension_group_mapping
                    ~f:(fun (_group_ext, platforms) ->
                      Base.List.exists platforms ~f:(fun p -> "." ^ p = ext)
                  )
                with
                | None ->
                  ( SSet.add ext unconditional_extensions,
                    grouped_extensions_with_conditional_extensions
                  )
                | Some (group_ext, _) ->
                  ( unconditional_extensions,
                    SMap.adjust
                      group_ext
                      (function
                        | None -> SSet.singleton ext
                        | Some s -> SSet.add ext s)
                      grouped_extensions_with_conditional_extensions
                  )
            )
          in
          let implementation_mref_of_platform_extension platform_ext = "./" ^ base ^ platform_ext in
          Some
            ( unconditional_extensions
              |> SSet.elements
              |> List.map implementation_mref_of_platform_extension,
              grouped_extensions_with_conditional_extensions
              |> SMap.elements
              |> List.map (fun (group_ext, conditional_exts) ->
                     ( implementation_mref_of_platform_extension group_ext,
                       conditional_exts
                       |> SSet.elements
                       |> List.map implementation_mref_of_platform_extension
                     )
                 )
            )
        else
          None
    )
  else
    None

File: rust_port/crates/flow_common/src/platform_set.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::collections::BTreeSet;
use std::path::Path;

use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::file_key::FileKey;

use crate::bitset::Bitset;
use crate::files;
use crate::files::FileOptions;
use crate::flow_projects::FlowProjects;
use crate::flow_projects::ProjectsOptions;

/// Platform set is represented as a bitset where each bit corresponds to a platform
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct PlatformSet(Bitset);

impl PlatformSet {
    pub fn is_subset(&self, other: &Self) -> bool {
        self.0.is_subset(&other.0)
    }

    pub fn no_overlap(&self, other: &Self) -> bool {
        self.0.no_overlap(&other.0)
    }

    pub fn to_platform_string_set(&self, file_options: &FileOptions) -> BTreeSet<FlowSmolStr> {
        file_options
            .multi_platform_extensions
            .iter()
            .enumerate()
            .filter_map(|(i, ext)| {
                if self.0.mem(i) {
                    Some(FlowSmolStr::new(ext.strip_prefix('.').unwrap()))
                } else {
                    None
                }
            })
            .collect()
    }
}

fn available_platforms_to_bitset(
    multi_platform_extensions: &[FlowSmolStr],
    available_platforms: &[FlowSmolStr],
) -> PlatformSet {
    let mut platform_set = Bitset::all_zero(multi_platform_extensions.len());
    for (i, ext) in multi_platform_extensions.iter().enumerate() {
        let platform = FlowSmolStr::new(ext.strip_prefix('.').unwrap_or(ext));
        if available_platforms.contains(&platform) {
            platform_set = platform_set.set(i)
        }
    }
    PlatformSet(platform_set)
}

pub fn available_platforms(
    file_options: &FileOptions,
    projects_options: &ProjectsOptions,
    filename: &str,
    explicit_available_platforms: Option<&[FlowSmolStr]>,
) -> Option<PlatformSet> {
    if !file_options.multi_platform {
        return None;
    }
    let multi_platform_extensions = &file_options.multi_platform_extensions;
    if let Some(result) =
        files::platform_specific_extensions_and_indices_opt(file_options, filename)
    {
        let mut platform_set = Bitset::all_zero(multi_platform_extensions.len());
        for (i, _) in result {
            platform_set = platform_set.set(i);
        }
        return Some(PlatformSet(platform_set));
    }

    match explicit_available_platforms {
        Some(explicit_platforms) => Some(available_platforms_to_bitset(
            multi_platform_extensions,
            explicit_platforms,
        )),
        None => {
            if let Some(supported_platforms_from_project_overrides) =
                FlowProjects::from_path(projects_options, filename).and_then(|projects| {
                    projects_options.multi_platform_ambient_supports_platform_for_project(projects)
                })
            {
                Some(available_platforms_to_bitset(
                    multi_platform_extensions,
                    &supported_platforms_from_project_overrides,
                ))
            } else {
                Some(PlatformSet(Bitset::all_one(
                    multi_platform_extensions.len(),
                )))
            }
        }
    }
}

/// Return a partitioned list of platform specific implementation mrefs to check existence,
/// of the form (unconditional_extensions, grouped_extensions_with_conditional_extensions).
///
/// - Every mref in unconditional_extensions must always exist.
/// - grouped_extensions_with_conditional_extensions is a list of (group_ext_mref, platform_ext_mref).
///   We first check if group_ext_mref exists. If that fails then platform_ext_mref must exist.
pub fn platform_specific_implementation_mrefs_of_possibly_interface_file(
    file_options: &FileOptions,
    platform_set: Option<&PlatformSet>,
    file: &FileKey,
) -> Option<(Vec<String>, Vec<(String, Vec<String>)>)> {
    if !file_options.multi_platform || !files::has_flow_ext(file) {
        return None;
    }
    let file = files::chop_flow_ext(file);
    let platform_set = platform_set.unwrap();
    file_options
        .module_file_exts
        .iter()
        .find_map(|module_file_ext| {
            if !file.check_suffix(module_file_ext) {
                return None;
            }
            let file_without_module_file_ext = file.chop_suffix(module_file_ext);
            let base = Path::new(file_without_module_file_ext.as_str())
                .file_name()
                .unwrap()
                .to_str()
                .unwrap();
            let platform_extensions: Vec<&FlowSmolStr> = file_options
                .multi_platform_extensions
                .iter()
                .enumerate()
                .filter_map(|(i, platform_ext)| {
                    if platform_set.0.mem(i) {
                        Some(platform_ext)
                    } else {
                        None
                    }
                })
                .collect();

            let mut unconditional_extensions: BTreeSet<&FlowSmolStr> = BTreeSet::new();
            let mut grouped_extensions_with_conditional_extensions: BTreeMap<
                &FlowSmolStr,
                BTreeSet<&FlowSmolStr>,
            > = BTreeMap::new();
            for ext in platform_extensions {
                match file_options
                    .multi_platform_extension_group_mapping
                    .iter()
                    .find(|(_, platforms)| {
                        platforms
                            .iter()
                            .any(|p| Some(p.as_str()) == ext.strip_prefix("."))
                    }) {
                    None => {
                        unconditional_extensions.insert(ext);
                    }
                    Some((group_ext, _)) => {
                        grouped_extensions_with_conditional_extensions
                            .entry(group_ext)
                            .or_insert_with(BTreeSet::new)
                            .insert(ext);
                    }
                }
            }

            let implementation_mref_of_platform_extension =
                |platform_ext: &str| format!("./{}{}", base, platform_ext);
            let unconditional_list: Vec<String> = unconditional_extensions
                .into_iter()
                .map(|ext| implementation_mref_of_platform_extension(ext))
                .collect();

            let grouped_list: Vec<(String, Vec<String>)> =
                grouped_extensions_with_conditional_extensions
                    .into_iter()
                    .map(|(group_ext, conditional_exts)| {
                        let group_mref = implementation_mref_of_platform_extension(group_ext);
                        let conditional_mrefs: Vec<String> = conditional_exts
                            .into_iter()
                            .map(|ext| implementation_mref_of_platform_extension(ext))
                            .collect();
                        (group_mref, conditional_mrefs)
                    })
                    .collect();

            Some((unconditional_list, grouped_list))
        })
}
